libs/trustchain/src/sdk.ts

```diff
@@ -309,6 +295,157 @@ export class SDK implements TrustchainSDK {
   }
 }
 
+type WithJwt = <T>(f: (jwt: JWT) => Promise<T>) => Promise<T>;
+
+type JwtExpirationCheck = {
+  hasExpired: boolean;
+  canBeRefreshed: boolean;
+};
+
+function networkCheckJwtExpiration(error: unknown): JwtExpirationCheck {
+  let hasExpired = false;
+  let canBeRefreshed = false;
+  // this assume live-network is used and we adapt to its error's format
+  if (error instanceof LedgerAPI4xx && error.message.includes("JWT is expired")) {
+    hasExpired = true;
+    canBeRefreshed = error.message.includes("/refresh");
+  }
+  return { hasExpired, canBeRefreshed };
+}
+
+async function genericWithJWT<T>(
+  f: (jwt: JWT) => Promise<T>,
+  initialJWT: JWT | undefined,
+  auth: () => Promise<JWT>,
+  policy: AuthCachePolicy = "cache",
+): Promise<T> {
+  function refresh(jwt: JWT) {
+    return api.refreshAuth(jwt).catch(e => {
+      log("trustchain", "JWT refresh failed, reauthenticating", e);
+      if (networkCheckJwtExpiration(e).hasExpired) {
+        return auth();
+      }
+      throw e;
+    });
+  }
```

Comment:
You can move it outside the function no?

Reply:
It's only a function to factorize the code of the two calls that contains logic specific to the general recovering of errors, not meant to be used externally also auth() is contextual here